WIP: Use CSS tooltips instead of Bootstrap

davwheat opened the pull request:
**Fixes flarum/issue-archive#131. Progresses flarum/issue-archive#179. Fixes flarum/issue-archive#144**

**Changes proposed in this pull request:**
This PR replaces Bootstrap tooltips with pure CSS tooltips. 🎉 

Despite this not implementing any extension-breaking API changes, some exts will need to modify their styling to remove uses of `overflow: hidden` to correctly show these CSS-only tooltips.

**Reviewers should focus on:**
Are there any issues with tooltip usage in core? **Issues would specifically relate to the use of `overflow: hidden` with tooltips-wielding nodes inside.**

**Known issues:**
- [x] Composer toolbar suffers from ["the problematic overflow spec"](https://css-tricks.com/popping-hidden-overflow/#the-problem)

**Screenshot**
![image](https://user-images.githubusercontent.com/7406822/111249340-791e8800-8603-11eb-8a21-d612ea5af9cb.png)
![image](https://user-images.githubusercontent.com/7406822/111249355-7de33c00-8603-11eb-8d58-0aeb5e47f32a.png)
![image](https://user-images.githubusercontent.com/7406822/111249369-83408680-8603-11eb-9f5e-7d4d0de33503.png)


**Confirmed**

- [x] Frontend changes: tested on a local Flarum installation.

davwheat committed:
Create Tooltip helper component

davwheat committed:
Create OS-based method of adding CSS workarounds

This adds a new HTML attribute `data-device-os` to the `body` element.

We can use this to perform OS-based workarounds for browser quirks and oddities.

We use this in this in the CSS-based tooltips to fix an issue on iOS where the :hover pseudoselector is not correctly triggered on non-interactive elements.

davwheat committed:
Add extra tooltip attributes

davwheat committed:
Add Tooltip component to common compat

davwheat committed:
Fix custom attrs being passed to DOM node

davwheat committed:
Format

davwheat committed:
Update tooltip styles

davwheat committed:
Add support for positioning (top, bottom, left, right)

davwheat committed:
Add basic Bootstrap tooltip backwards compatibility layer

davwheat committed:
Add hacky fix for CSS tooltips in the Composer

This fix involves making the original outer element taller than needed (90px), and will mean that the tooltips can be shown (as they will be within the 90px tall container).

The element then receives `display: flex` meaning that the controls themselves can be aligned to the end of the flex container, pushing them to the bottom of the container to where they are expected to be.

We also have to add a CSS custom property (https://caniuse.com/css-variables) using the data calculated in the Less stylesheet for height calculations when resizing the composer. This ensures that the text area height actually fills the entire Composer, rather than leaving a small gap at the bottom.

Additionally, we use a CSS background to replace where the border was originally to keep the styling consistent. This is achieved with some clever Less styling.

We also add a new event listener to watch for window resizes. This allows us to add a new `TextEditor-controls--hasScrollbar` class if the toolbar has a scrollbar. This allows us to fix weird styling issues surrounding this hack for when the toolbar has a horizontal scrollbar by reducing the bottom padding.

Unfortunately this hack can't fix the tooltips on elements close to the edges of the screen -- there's no way to set a minimum-left in CSS. Should we maybe consider *some* JS to fix this, or live with it instead? See: https://medium.com/carwow-product-engineering/building-a-simple-tooltip-component-that-never-goes-off-screen-c7039dcab5f9

davwheat committed:
Add padding between top of composer and load more button

davwheat committed:
Replace deprecated jQuery calls

davwheat committed:
Stretch Composer toolbar to edges on mobile

davwheat committed:
Replace OS check with @media (any-hover)

davwheat committed:
Forgot to remove import

askvortsov1 reviewed (changes requested):
I don't mind the hack that much, but going off-screen is problematic IMO. Would JS even be able to fix this if we're using pseudoelements?

If we need a placement library, https://github.com/tobyzerner/placement.js could be a candidate

davwheat committed:
Update JSDoc comments and types

davwheat committed:
Set show on focus by default

This wasn't done before! Oops

davwheat committed:
Add fake destroy method

This may be called to destroy a tooltip to recreate it with new content. This is unnecessary using the CSS method as the tooltip content is read dynamically.

davwheat committed:
Add limitations section to back-compat layer JSDoc

davwheat committed:
Use cleaner fallback method

davwheat committed:
Add WIP MutationObserver for watching for new/modified tooltips

SychO9 commented:
Whoa there's a lot going on here, thanks for the hard work 🤗 

I like replacing JS with CSS where possible but I have mixed feelings about this change, having the tooltip be a direct descendant of the element we're tooltiping is something I am *uncomfortable with*/*uncertain about* since the only scenarios we're faced with are the default ones. I see we've added a hack to fix the overflow issue where it occurs in core, but an extension facing the same situation will have to do something similar, in addition to potential scenarios where z-indexes might cause trouble.

I don't have any concrete examples though which is not helpful to anyone but I wanted to voice my concern nonetheless.
Because I am uncertain about this 😕 I will refrain from reviewing.

askvortsov1 commented:
> Because I am uncertain about this 😕 I will refrain from reviewing.

On the contrary, I think this is exactly a reason to review. I also share some of these concerns. This is all quite cool, but it's also a huge PR with a lot of edge cases and workarounds. Having CSS only tool tips would be ideal, but if it's not possible without this much modification and hacks, I would prefer to wait. Our goal currently is to:

- get rid of bootstrap stuff (possibly replaced with a smaller library)
- provide an abstract API so we can change how tool tips are implemented in flarum without breaking use in extensions.

I don't think I'm comfortable with introducing so many workarounds, especially directly in flarum source code. Also, if we need so many workarounds, is it really CSS-only tool tips? I know you've put a huge amount of time and work into this, but I'm not sure we should proceed with this solution right now (maybe in some time browser support will be better?).

davwheat committed:
More WIP code

Code created before discussion to leave for 1.1+
